addons/sale/models/payment_transaction.py

```diff
@@ -49,32 +46,37 @@ def _set_pending(self, state_message=None):
             sales_orders._send_order_confirmation_mail()
 
     def _check_amount_and_confirm_order(self):
-        self.ensure_one()
-        for order in self.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent')):
-            if order.currency_id.compare_amounts(self.amount, order.amount_total) == 0:
-                order.with_context(send_email=True).action_confirm()
-            else:
-                _logger.warning(
-                    '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %r, got %r',
-                    self.acquirer_id.provider,order.name, order.id,
-                    order.amount_total, self.amount,
-                )
-                order.message_post(
-                    subject=_("Amount Mismatch (%s)", self.acquirer_id.provider),
-                    body=_("The order was not confirmed despite response from the acquirer (%s): order total is %r but acquirer replied with %r.") % (
-                        self.acquirer_id.provider,
-                        order.amount_total,
-                        self.amount,
+        """ Confirm the sale order based on the amount of a transaction.
+
+        :return: confirmed sale orders
+        :rtype: a set of sale orders
+        """
+        sale_orders_confirmed = self.env['sale.order']
+        for tx in self:
+            # We only support the flow where exactly one quotation is linked to a transaction and
+            # vice versa.
+            order = tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent'))
+            if len(tx.sale_order_ids) == 1 and order and len(order.transaction_ids) == 1:
+                # Check if the SO is fully paid
+                if order.currency_id.compare_amounts(tx.amount, order.amount_total) == 0:
+                    order.with_context(send_email=True).action_confirm()
+                    sale_orders_confirmed |= order
+                else:
+                    _logger.warning(
+                        '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %s, got %s',
+                        tx.provider,
+                        order.name,
+                        order.id,
+                        format_amount(order.env, order.amount_total, order.currency_id),
+                        format_amount(tx.env, tx.amount, tx.currency_id),
                     )
-                )
+        return sale_orders_confirmed
 
     def _set_authorized(self, state_message=None):
         """ Override of payment to confirm the quotations automatically. """
         super()._set_authorized(state_message=state_message)
         sales_orders = self.mapped('sale_order_ids').filtered(lambda so: so.state in ('draft', 'sent'))
-        for tx in self:
-            tx._check_amount_and_confirm_order()
-
+        self._check_amount_and_confirm_order()
         # send order confirmation mail
         sales_orders._send_order_confirmation_mail()
```

Comment:
```suggestion
        confirmed_orders = self._check_amount_and_confirm_order()
        # send order confirmation mail
        confirmed_orders._send_order_confirmation_mail()
```
(and remove `sales_orders` variable)

Reply:
@AntoineVDV was clear that we should not change the way sending mail works in this task. If the mail is not sent twice, it's not a bug. Although, I agree that the mailing system is wobbly and should be improved.

Reply:
![image](https://user-images.githubusercontent.com/19364378/149190600-9c792a64-5237-4133-9ecf-3ef4c635f4d5.png)
(on your branch, with a little tweak to support authorization with "Test")

Reply:
Shit. I though I was getting away with this.I'll look into it tomorrow :p